### DISCUSSION TITLE
RFC: Collections, ML Workflow Phases, and Inspection Certification

### DISCUSSION THREAD

**Discussion by garland3** [sandialabs/VISTA#58] [Ideas]
## Problem

VISTA currently organizes images in a flat list per project. There is no way to:

- Group images into meaningful subsets (a batch, a lot, a training set)
- Define labels/classes scoped to a specific task rather than the whole project
- Lock a subset of work as finalized while other work continues
- Track the lifecycle of annotation work through stages (labeling, review, sign-off)
- Connect training data to the ML model it produced to the inference results it generated

For industrial inspection use cases, we need to certify that a lot was fully reviewed by humans and then lock it as the permanent record.

### What is a Collection?

A **collection** is a named group of images within a project. It is the primary unit of work.

```
Project
  +-- Collection A ("Training Set v1")
  |     +-- 500 images
  |     +-- Label classes: crack, void, delamination
  |     +-- Bounding box annotations
  +-- Collection B ("Lot 2024-0847 Inspection")
  |     +-- 1000 images
  |     +-- Label classes: defect, acceptable
  |     +-- ML predictions + human corrections
  +-- Collection C  ...
```

**Key properties:**
- An image can belong to multiple collections
- Label classes (both whole-image and bounding box) are scoped per collection, not per project
- Each collection has its own lifecycle state independent of other collections
- Collections track who created them, when, and all modifications via audit trail

### Data Model

```
ImageCollection:
  id, project_id, name, description
  purpose: "labeling" | "inference"
  phase: "draft" | "annotating" | "review" | "certified"
  source_collection_id: FK (nullable, lineage)
  ml_model_name, ml_model_version (nullable, for inference collections)
  certified_by_id, certified_at, certification_notes (nullable)
  archived_at (nullable, housekeeping flag)
  created_by_id, created_at, updated_at

CollectionMembership:
  collection_id, image_id, added_by_id, added_at
  (unique constraint: one image per collection)

ImageClass:
  + collection_id FK (scoped to collection, not project)

BoundingBoxClass:
  + collection_id FK (scoped to collection, not project)
```

## Collection Purpose

Set at creation, never changes:

| Purpose | Description |
|---|---|
| `labeling` | Humans annotate images from scratch (building training data) |
| `inference` | ML model ran predictions on these images; humans review and correct |

## Collection Phases (Lifecycle)

```
draft --> annotating --> review --> certified
                           ^           |
                           +--- reopen-+  (requires reason, creates audit event)
```

| Phase | Images | Annotations | Who | What happens |
|---|---|---|---|---|
| `draft` | Add/remove freely | Define classes | Setup person | Build the image set |
| `annotating` | Frozen roster | Create annotations | ML pipeline or humans | Initial labeling pass |
| `review` | Frozen | Edit/add/remove | Human reviewers | Correct and verify annotations |
| `certified` | Frozen | Frozen | Certifier signs off | Locked permanent record |

Phase transitions enforce rules:
- `draft` to `annotating`: freezes the image roster (no add/remove)
- `annotating` to `review`: no new annotations from ML; only human review edits
- `review` to `certified`: requires all images reviewed, no flagged annotations
- After certification: HTTP 423 on all mutations

## Review Phase: Human-in-the-Loop Editing

The review phase is not just approve/reject. Reviewers can actively correct the work:

- **Edit** a bounding box (adjust coordinates, change class)
- **Delete** a false positive
- **Add** a missed detection
- **Mark each image as reviewed** -- explicit per-image sign-off

### Per-Image Review Tracking

For industrial inspection, you must prove every image was seen by a human:

```
ImageReview:
  id, collection_id, image_id
  reviewer_id, status ("reviewed" | "flagged")
  notes, reviewed_at
```

Certification requires every image to have an `ImageReview` with status `reviewed`.

### Annotation Provenance

Track where each annotation came from:

```
UserAnnotation:
  + origin: "manual" | "ml_prediction" | "ml_corrected"
  + original_annotation_id: FK (nullable, points to the ML prediction that was corrected)
```

When a reviewer edits an ML prediction, the original is preserved and a new `ml_corrected` annotation is created pointing back. This lets you measure ML accuracy over time.

## Certification

Certification is the deliverable. The locked collection IS the inspection record.

Pre-conditions enforced:
- Every image has a `reviewed` ImageReview (no gaps)
- No annotations in `flagged` state
- At least one reviewer touched every image

After certification, the collection records who certified, when, and any notes (e.g., "Lot 2024-0847 reviewed per SOP-123").

## Lineage: Connecting Training to Inference

An inference collection tracks where it came from:
- `source_collection_id` points to the certified labeling collection that trained the model
- `ml_model_name` and `ml_model_version` record which model produced the predictions

This answers: given these inference results, what training data and model version produced them?

## Full Workflow Example

```
1. Create labeling collection (phase: draft)
   - Add reference images, define defect classes (crack, void, etc.)
   - Move to annotating: humans draw bounding boxes on training images
   - Move to review: QA pass on the labels
   - Certify: frozen ground truth dataset

2. Export certified collection, train ML model externally

3. Create inference collection (phase: draft)
   - source_collection_id = the certified labeling collection
   - ml_model = "defect-detector-v3"
   - Add production lot images
   - Move to annotating: ML pipeline writes predictions via API

4. Move to review
   - Inspectors correct ML predictions (edit/add/remove bboxes)
   - Each inspector marks images as reviewed
   - Progress dashboard: "847/1000 images reviewed"

5. Certify
   - System enforces: all images reviewed, no flagged items
   - Inspector lead signs off
   - Collection locked permanently
   - "Lot 2024-0847: 1000 images, 23 defects found, certified by J. Smith"
```

## Open Questions

- Do we need role-based restrictions on who can certify vs. who can review?
- Should reopening a certified collection require a specific permission level?
- Export formats for training data: COCO, YOLO, Pascal VOC, or all three?
- Should there be a bulk "import predictions" endpoint for ML pipelines, or use existing per-annotation API?

## Related

- Issue #56 tracks the foundational implementation (collections, bounding box annotations, review workflow, audit trail)
- Prototype exists on `feature/collections-primary` branch (not yet merged to main)

**Comment by garland3** [comment]
**Context:** This RFC supersedes issue #56, which covered the initial implementation of bounding box annotations, collections, review workflow, and audit trail. That foundational work was merged in PR #53. This discussion picks up where #56 left off, focusing on the collection lifecycle phases needed for ML training/inference workflows and industrial inspection certification.

**Comment by garland3** [comment]
**Correction:** The previous comment was inaccurate. The collections, bounding box annotations, review workflow, and audit trail from issue #56 have **not** been merged to main. That work exists only on the `feature/collections-primary` branch. Issue #56 remains open.

This RFC builds on and extends that unmerged work. The implementation plan should include both the foundational collections work from #56 and the lifecycle/certification extensions proposed here.

**Comment by garland3** [comment]
<img width="1536" height="1024" alt="image" src="https://github.com/user-attachments/assets/138a011d-5a9a-41f3-827a-4c3eb826fd6e" />
